platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java

```diff
@@ -60,24 +61,22 @@
 public abstract class Ear extends Jar {
     public static final String EAR_EXTENSION = "ear";
 
-    private String libDirName;
     private final Property<Boolean> generateDeploymentDescriptor;
     private DeploymentDescriptor deploymentDescriptor;
-    private CopySpec lib;
+    private final CopySpec lib;
     private final DirectoryProperty appDir;
 
     public Ear() {
         getArchiveExtension().set(EAR_EXTENSION);
         setMetadataCharset("UTF-8");
         generateDeploymentDescriptor = getObjectFactory().property(Boolean.class);
         generateDeploymentDescriptor.convention(true);
-        lib = getRootSpec().addChildBeforeSpec(getMainSpec()).into(
-            callable(() -> GUtil.elvis(getLibDirName(), DEFAULT_LIB_DIR_NAME))
-        );
+        lib = getRootSpec().addChildBeforeSpec(getMainSpec()).into(getLibDirName().orElse(DEFAULT_LIB_DIR_NAME));
+        AtomicReference<SetProperty<EarModule>> topLevelModules = new AtomicReference<>(getObjectFactory().setProperty(EarModule.class));
         getMainSpec().appendCachingSafeCopyAction(action(details -> {
             if (generateDeploymentDescriptor.get()) {
                 checkIfShouldGenerateDeploymentDescriptor(details);
-                recordTopLevelModules(details);
+                topLevelModules.set(union(topLevelModules.get(), recordTopLevelModules(details)));
```

Comment:
This only needs to be a `Provider<Set>`, not a `SetProperty` necessarily. Can we use `.zip()` here instead? Or can we just use `Set` and union that? I don't understand why we are using `Property` types here at all.

Reply:
> Or can we just use Set and union that?

We now use a Set